src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+            }
+
+            return result;
+        }
+
+        private void AddToDecisionTree(DecisionTree decisionTree, BoundPattern pattern, BoundExpression guard, BoundPatternSwitchLabel label)
+        {
+            if (guard?.ConstantValue == ConstantValue.False) return;
+            switch (pattern.Kind)
+            {
+                case BoundKind.ConstantPattern:
+                    {
+                        var constantPattern = (BoundConstantPattern)pattern;
+                        AddByValue(decisionTree, constantPattern.Value, (e, t) => new DecisionTree.Guarded(e, t, default(ImmutableArray<KeyValuePair<BoundExpression, LocalSymbol>>), guard, label), label.HasErrors);
+                        break;
+                    }
+                case BoundKind.DeclarationPattern:
+                    {
+                        var declarationPattern = (BoundDeclarationPattern)pattern;
+                        DecisionMaker maker = (e, t) => new DecisionTree.Guarded(e, t, ImmutableArray.Create(new KeyValuePair<BoundExpression, LocalSymbol>(e, declarationPattern.LocalSymbol)), guard, label);
+                        if (declarationPattern.IsVar)
+                        {
+                            Add(decisionTree, maker);
+                        }
+                        else
+                        {
+                            AddByType(decisionTree, declarationPattern.DeclaredType.Type, maker);
+                        }
+                        break;
+                    }
+                case BoundKind.WildcardPattern:
+                    // We do not yet support a wildcard pattern syntax. It is used exclusively
+                    // to model the "default:" case, which is handled specially in the caller.
+                    throw ExceptionUtilities.UnexpectedValue(pattern.Kind);
+            }
+        }
+
+        private ErrorCode CheckSubsumed(BoundPattern pattern, DecisionTree decisionTree, bool inputCouldBeNull)
+        {
+            if (decisionTree.MatchIsComplete) return ErrorCode.ERR_PatternIsSubsumed;
+            switch (pattern.Kind)
+            {
+                case BoundKind.ConstantPattern:
+                    {
+                        var constantPattern = (BoundConstantPattern)pattern;
+                        if (constantPattern.Value.HasErrors || constantPattern.Value.ConstantValue == null)
+                        {
+                            // since this will have been reported earlier, we use ErrorCode.ERR_NoImplicitConvCast
+                            // as a flag to suppress errors in subsumption analysis.
+                            return ErrorCode.ERR_NoImplicitConvCast;
+                        }
+                        bool isNull = constantPattern.Value.ConstantValue.IsNull;
+                        if (isNull && !inputCouldBeNull) return ErrorCode.ERR_PatternIsSubsumed;
```

Comment:
I am not sure I am following this logic. If the pattern is constant null and the input is not null, then why the input is subsumed? Could you add an explanation as a comment?